types/src/ledger_info.rs

```diff
@@ -373,6 +374,192 @@ impl LedgerInfoWithPartialSignatures {
+    }
+
+    // Collecting all the authors from verified signatures, unverified signatures and the aggregated signature.
+    pub fn all_voters(&self, verifier: &ValidatorVerifier) -> Vec<AccountAddress> {
+        let authors = self.verified_aggregate_signature.get_signers_addresses(
+            verifier
+                .get_ordered_account_addresses_iter()
+                .collect_vec()
+                .as_slice(),
+        );
+        let authors = authors
+            .iter()
+            .chain(self.verified_signatures.signatures().keys());
+        let authors = authors
+            .chain(self.unverified_signatures.keys())
+            .collect::<HashSet<_>>();
+        authors.iter().cloned().cloned().collect()
+    }
+
+    pub fn check_voting_power(
+        &self,
+        verifier: &ValidatorVerifier,
+    ) -> std::result::Result<u128, VerifyError> {
+        let all_voters = self.all_voters(verifier);
+        verifier.check_voting_power(all_voters.iter().collect_vec().into_iter(), true)
+    }
+
+    pub fn aggregate_and_verify(
+        &mut self,
+        verifier: &ValidatorVerifier,
+    ) -> Result<LedgerInfoWithSignatures, VerifyError> {
+        self.check_voting_power(verifier)?;
+
+        // Compute signatures from verified signatures and unverified signatures that are not already
+        // included in the verified aggregate signature.
+        let mut partial_signatures = self.verified_signatures.clone();
+        for (author, signatures) in self.unverified_signatures.iter() {
+            partial_signatures.add_signature(*author, signatures[0].clone());
+        }
+        let authors_in_aggregated_sig = self.verified_aggregate_signature.get_signers_addresses(
+            verifier
+                .get_ordered_account_addresses_iter()
+                .collect_vec()
+                .as_slice(),
+        );
+        for author in authors_in_aggregated_sig {
+            partial_signatures.remove_signature(author);
+        }
+
+        if partial_signatures.is_empty() {
+            return Ok(LedgerInfoWithSignatures::new(
+                self.ledger_info.clone(),
+                self.verified_aggregate_signature.clone(),
+            ));
+        }
+
+        let aggregated_sig = verifier.aggregate_new_signatures_with_existing(
+            &partial_signatures,
+            Some(self.verified_aggregate_signature.clone()),
+        )?;
+        let verified_aggregate_signature =
+            match verifier.verify_multi_signatures(self.ledger_info(), &aggregated_sig) {
+                Ok(_) => aggregated_sig,
+                Err(_) => {
+                    let verified = self
+                        .unverified_signatures
+                        .clone()
+                        .into_par_iter()
+                        .flat_map(|(account_address, signatures)| {
+                            // Get the first signature that verifies correctly.
+                            for signature in signatures {
```

Comment:
not sure I understand why this is a vector of signatures

Reply:
As these are unverified signatures, in case a malicious validator sends multiple votes, I'm storing all those signatures in a vector.
Should I remove this and store just the first vote sent by a validator?

Reply:
we probably should keep the last one